tamagotchi/io/src/lib.rs

```diff
@@ -13,20 +15,162 @@ impl Metadata for TamagotchiMetadata{
+        TamagotchiState::verify_limit(&mut self.rested);
+        self.rested_block = current_block_height;
+        self.rested += FILL_PER_SLEEP
+            .saturating_sub(ENERGY_PER_BLOCK * (current_block_height - self.rested_block));
+        TamagotchiState::verify_limit(&mut self.rested);
+    }
+
+    pub fn verify_limit(mood_param: &mut u64) {
+        if *mood_param > MAX_MOOD_VALUE {
+            *mood_param = MAX_MOOD_VALUE;
+        }
+        if *mood_param < MIN_MOOD_VALUE {
+            *mood_param = MIN_MOOD_VALUE;
+        }
+    }
+
+    pub fn verify_ownership(&self, source: ActorId) -> bool {
+        self.owner == source
+    }
+
+    pub fn verify_allowed_account(&self, source: ActorId) -> bool {
+        match self.allowed_account {
+            Some(account) => account == source,
+            None => false,
+        }
+    }
+
+    pub fn verify_permission(&self, source: ActorId) -> bool {
+        self.verify_ownership(source) || self.verify_allowed_account(source)
+    }
+
+    pub async fn approve_tokens(&mut self, account: &ActorId, amount: u128) -> TmEvent {
+        let (transaction_id, account, amount) = if let Some((
+            ft_transaction_id,
+            prev_account,
+            prev_amount,
+        )) = self.approve_transaction
+        {
+            if prev_account != *account || prev_amount != amount {
+                panic!("Please complete the previous transaction");
+            } else {
+                (ft_transaction_id, prev_account, prev_amount)
+            }
+        } else {
+            let ft_transaction_id = self.transaction_id;
+            self.transaction_id = self.transaction_id.wrapping_add(1);
+            self.approve_transaction = Some((ft_transaction_id, *account, amount));
+            (ft_transaction_id, *account, amount)
+        };
+        if let Some(contract) = self.ft_contract {
+            debug!("Sending approve tokens message to FT contract");
+            let result = msg::send_for_reply_as::<_, FTokenEvent>(
+                contract,
+                FTokenAction::Message {
+                    transaction_id,
+                    payload: LogicAction::Approve {
+                        approved_account: account,
+                        amount,
+                    },
+                },
+                0,
+                0,
+            )
+            .expect("Error sending approve tokens message")
+            .await;
+            self.approve_transaction = None;
+            match result {
+                Ok(_) => return TmEvent::Approve(account),
+                Err(_) => return TmEvent::ApprovalError,
+            }
+        } else {
+            debug!("FT contract not set");
+            panic!("FT contract not set");
+        };
+    }
 }
 
 #[derive(Encode, Decode, TypeInfo)]
-pub enum TmAction{
+pub enum TmAction {
     Name,
-    Age
+    Age,
+    Feed,
+    Sleep,
+    Play,
+    Transfer(ActorId),
+    Approve(ActorId),
+    RevokeApproval,
```

Comment:
No need to include previous changes to the PR; they are to be merged into the master branch.